### PR TITLE
Add Windows JACK support

### LINKED ISSUE #2136: Jack audio support on windows
I don't know if this is the proper place for a feature request like this, but eventually it would be cool to have [jack](http://jackaudio.org/) support on windows. There are obviously much more pressing things to work on though.

### PR DESCRIPTION
**Pull request by tresf**
JACK is surprisingly plug-and-play with regards to LMMS on Windows, thanks to the hard work of @falkTX and the rest of the JACK2 team combined with our already existing `weakjack` (thanks @x42) module (preventing the need for windows dev libs).

* Adds JACK build support for Windows systems
* Removes libjack-dev dependency unless specified
* Adds jack2 as a submodule, related #3963 
* Bumps weakjack submodule to add MSVC support.

To test:

* Install JACK for Windows from the website
* Launch JACK with ASIO support per JACK's website instructions
* Switch LMMS backend to JACK
* Configure using `qjackctl`, found in JACK install directory

image

Closes #2136

### CODE REVIEW

**Comment by PhysSong** [comment]
The `weakjack` submodule should be updated to support MSVC, see https://github.com/x42/weakjack/pull/9.

File: src/3rdparty/weakjack/CMakeLists.txt [hunk 1]
```diff
@@ -7,6 +7,6 @@ IF(LMMS_HAVE_WEAKJACK)
 	
 	# Library stub for AppImages running on systems without jack
 	ADD_LIBRARY(weakjack MODULE weakjack/weak_libjack.c weakjack/weak_libjack.h)
-	TARGET_INCLUDE_DIRECTORIES(weakjack PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/weakjack ${JACK_INCLUDE_DIRS})
+	TARGET_INCLUDE_DIRECTORIES(weakjack PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/weakjack ${CMAKE_CURRENT_SOURCE_DIR}/../jack2/common ${JACK_INCLUDE_DIRS})
```
Comment (PhysSong) [inline src/3rdparty/weakjack/CMakeLists.txt:10]: Could you explain why did you add this?

Comment (falkTX) [reply src/3rdparty/weakjack/CMakeLists.txt:10]: because using weakjack the real jack is not needed, only the header files.

but in that case the right submodule is actually https://github.com/jackaudio/headers

Comment (x42) [reply src/3rdparty/weakjack/CMakeLists.txt:10]: For Ardour windows builds, we use http://ardour.org/files/deps/jack_win3264.tar.xz

It's a re-zip of jack headers, .def, .lib, and  32/64bit .dll  from jackd 1.9.10 with a pkg-config file added for convenience,
The files were made using sletz' last release of jack for windows back in 2012:
* https://dl.dropboxusercontent.com/u/28869550/Jack_v1.9.10_32_setup.exe
* https://dl.dropboxusercontent.com/u/28869550/Jack_v1.9.10_64_setup.exe

With weakjack one only needs the headers, but the other files come in handy for unit-tests.

Comment (tresf) [reply src/3rdparty/weakjack/CMakeLists.txt:10]: > Could you explain why did you add this?

This part seems answered.

> the right submodule is actually https://github.com/jackaudio/headers

Done via b852863, but for the record, now the submodule folder is mismatched from the project name and this is inconsistent with the [other submodules we clone](https://github.com/LMMS/lmms/issues/3963) due to needing a `jack/` prefix which `jack2` was already providing but I digress. :)

> With `weakjack` one only needs the headers, but the other files come in handy for unit-tests.

We're moving most of our Windows system to use `vcpkg` bootstrapping and building on-the-fly.  I don't love this strategy from a time-to-build-perspective, but it certainly avoids the pre-built fetching on Windows, which in 2020 still seems to lack any consistent dependency system.  For the scope of this PR this is all moot since we don't have JACK unit tests (yet) but I'm sure this will come back up in the future.  Anyway, thanks kindly for sharing.  I had initially planned on making a [viable dependency system using precompiled binaries and scoop.sh](https://github.com/lukesampson/scoop-extras/pulls?q=is%3Apr+author%3Atresf+is%3Aclosed), but never came to fruition.  For now, `vcpkg` seems to be the least cluster**** solution.  I mention all of this because as a project, we're always interested in how others do it.  LMMS has had a rocky history of building on Windows and we greatly appreciate the info.

Comment (tresf) [reply src/3rdparty/weakjack/CMakeLists.txt:10]: @falkTX the recommended header repo throws the following error in AppVeyor:

```
C:\projects\lmms\src\3rdparty\jackaudio\jack/types.h(25): fatal error C1083: Cannot open include file: 'pthread.h':
```

Which seems to be referenced here:
https://github.com/jackaudio/headers/blob/master/types.h

... but not here:
https://github.com/jackaudio/jack2/blob/develop/common/jack/types.h

Is this just a matter of the headers repo needing updating?

Comment (tresf) [reply src/3rdparty/weakjack/CMakeLists.txt:10]: > > Could you explain why did you add this?

> This part seems answered.

As usually @PhysSong's right 🤦 , I had forgotten to delete the redundant include, the variable is already populated earlier.  Patching now...

Comment (tresf) [reply src/3rdparty/weakjack/CMakeLists.txt]: Done via 9c1cf61.  Moving other conversation points to another line-number for context.

File: .gitmodules [hunk 2]
```diff
@@ -40,3 +40,6 @@
 [submodule "plugins/carlabase/carla"]
 	path = plugins/carlabase/carla
 	url = https://github.com/falktx/carla
+[submodule "src/3rdparty/jackaudio/jack"]
```
Comment (tresf) [inline .gitmodules:43]: @falkTX wrote:
> the right submodule is actually https://github.com/jackaudio/headers

Done via b852863, however now AppVeyor throws the following:

```
C:\projects\lmms\src\3rdparty\jackaudio\jack/types.h(25): fatal error C1083: Cannot open include file: 'pthread.h':
```

Which seems to be referenced here:
https://github.com/jackaudio/headers/blob/master/types.h

... but not here:
https://github.com/jackaudio/jack2/blob/develop/common/jack/types.h

Is this just a matter of the headers repo needing updating?  I'll also bump weakjack for MSVC support incase that's related.

File: src/3rdparty/weakjack/CMakeLists.txt (hunk 1 above)
Comment (falkTX) [reply src/3rdparty/weakjack/CMakeLists.txt]: > [](https://github.com/jackaudio/jack2/blob/develop/common/jack/types.h)
> 
> Is this just a matter of the headers repo needing updating?

yeah, the headers at the moment target jack1.

**Comment by tresf** [comment]
> The `weakjack` submodule should be updated to support MSVC, see x42/weakjack#9.

Done via b1ab5ae.

File: src/3rdparty/weakjack/CMakeLists.txt (hunk 1 above)
Comment (tresf) [reply src/3rdparty/weakjack/CMakeLists.txt]: > yeah, the headers at the moment target jack1.

Should I revert then?

Comment (falkTX) [reply src/3rdparty/weakjack/CMakeLists.txt]: yes sorry.

File: .gitmodules (hunk 2 above)
Comment (tresf) [reply .gitmodules]: Per @falkTX, this needs to be reverted as the `headers` repo is based off of jack1, not jack2 and has issues compiling on MSVC.  Done via
f55f42d.

**Comment by tresf** [comment]
All notes have been addressed, but there's a lingering issue from #5636 preventing AppVeyor to succeed.  I'm fine waiting for that to be fixed to make sure AppVeyor is happy or merging as-is.  Thoughts welcome.   Pinging @Spekular since he authored  #5636.

**Comment by Spekular** [comment]
Huh. Is this related to the issue fixed by https://github.com/LMMS/lmms/pull/5678 somehow? Other builds since then have been fine, so I'm frankly not sure why an issue would pop up now.

**Comment by tresf** [comment]
> Huh. Is this related to the issue fixed by #5678 somehow? Other builds since then have been fine, so I'm frankly not sure why an issue would pop up now.

Ok, sorry for the ping, I did search beforehand. I'll fast-forward and it should fix it, thanks.

**Comment by tresf** [comment]
Ok, I fast-forwarded my branch didn't fix this @Spekular, can you comment on the AppVeyor error?  It's the same as before.

**Comment by PhysSong** [comment]
I think [`(std::numeric_limits<int>::max)()` may work](https://stackoverflow.com/a/2789509). I'll test it out and fix `lmms_basic.h` as well if that works.

**Comment by tresf** [comment]
> I'm frankly not sure why an issue would pop up now.

@Spekular as discussed in #devtalk on [Discord](https://lmms.io/chat), it appears this is as a result of adding `jack.h` into AppVeyor, which is new with this PR.  So, although the symptom is the same, the cause is likley a 3rdparty header we don't manage.  I've made my first attempt at patching this in d56db57.  If that doesn't work, I might fire up a MSVC environment instead.

> I think `(std::numeric_limits<int>::max)()` may work.

@DomClark shared his sentiments on Discord, and I tend to agree with them that we shouldn't have to worry about adding parens to each call to `max`.  

>  I'll test it out and fix `lmms_basic.h` as well if that works.

I started writing this on a [separate branch, but was largely unsuccessful](https://github.com/LMMS/lmms/compare/master...tresf:nominmax).  I'll revisit if needed.

File: plugins/vst_base/RemoteVstPlugin.cpp [hunk 3]
```diff
@@ -40,10 +40,6 @@
 
 #ifdef LMMS_BUILD_LINUX
 
-#ifndef NOMINMAX
-#define NOMINMAX
-#endif
-
```
Comment (PhysSong) [inline plugins/vst_base/RemoteVstPlugin.cpp:46]: Please keep this or modify relevant CMake files, we don't propagate all C++ flags when building `RemoteVstPlugin` using `ExternalProject`.

**Comment by tresf** [comment]
> I'll test it out and fix `lmms_basic.h` as well if that works.

I decided to opt for a top-level... 
```cmake
ADD_DEFINITIONS(-DNOMINMAX)
```
... which I believe solves the problem proactively and holistically.

On the contrary, `lmms_basics.h` can't extend to headers which we don't manage, like `weak_jack.h`, so I believe this to be ideal.

Components which build using `ExternalProject_Add` may not import all flags, so `RemoteVstPlugin` is left as-is.

**Comment by tresf** [comment]
Something's odd with Travis, it's not even queuing my builds (it's been hours and I've tried restarting it once).  I'm not holding up the PR for that specifically.  There's also a failure after rebase, but it's unrelated and explained here: https://github.com/LMMS/lmms/commit/435dbc5e00fa06cd789693a56e6b340f4c7722ab#r43742718.  Merging.
